### PR TITLE
added llvm 4.0.0 support

### PR DESCRIPTION
**Pull request by foerdi**

### CODE REVIEW

File: source/llvm/constants.d [hunk 1]
```diff
@@ -478,8 +481,15 @@ enum : llvm_lto_status
 }
 
 /+ LTO +/
-
-static if (LLVM_Version >= asVersion(3, 7, 0))
+static if (LLVM_Version >= asVersion(4, 0, 0))
+{
+	const uint LTO_API_VERSION = 21;
+}
+else static if (LLVM_Version >= asVersion(3, 9, 0))
```
Comment (Calrama) [inline source/llvm/constants.d]: Hm, seems like 3.9.0 support wasn't complete here.

**Comment by foerdi** [comment]
I updated the config.d file based on the files in https://github.com/llvm-mirror/llvm/tree/master/lib/Target
Is there a better location for this information?

**Comment by Calrama** [comment]
@foerdi llvm-d includes a [tool for that](https://github.com/Calrama/llvm-d/blob/master/tools/get-targets)
Invoke like so:
$ ./tools/get-targets llvm-4.0.src/

File: source/llvm/config.d [hunk 2]
```diff
@@ -126,7 +133,9 @@ immutable LLVM_AsmParsers = {
 
 /// LLVM Targets with Disassembler capability (if enabled)
 immutable LLVM_Disassemblers = {
-	       static if (LLVM_Version >= asVersion(3, 9, 0)) {
+	static if (LLVM_Version >= asVersion(3, 9, 0)) {
```
Comment (Calrama) [inline source/llvm/config.d]: Should've been asVersion(4, 0 ,0) here btw.

File: source/llvm/config.d [hunk 3]
```diff
@@ -55,32 +56,36 @@ immutable LLVM_Version = asVersion(LLVM_VERSION_MAJOR, LLVM_VERSION_MINOR, LLVM_
...
-		} else static if (LLVM_Version >= asVersion(3, 3, 0)) {
-			return ["AArch64","ARM","CppBackend","Hexagon","MBlaze","MSP430","Mips","NVPTX","PowerPC","R600","Sparc","SystemZ","X86","XCore"];
-		} else static if (LLVM_Version >= asVersion(3, 2, 0)) {
-			return ["ARM","CellSPU","CppBackend","Hexagon","MBlaze","MSP430","Mips","NVPTX","PTX","PowerPC","Sparc","X86","XCore"];
-		} else {
-			return ["ARM","CellSPU","CppBackend","Hexagon","MBlaze","MSP430","Mips","PTX","PowerPC","Sparc","X86","XCore"];
-		}
-	}().map!(t => "version (LLVM_Target_" ~ t ~ ") targets ~= \"" ~ t ~ "\";").joiner.array);
+			static if (LLVM_Version >= asVersion(4, 0, 0)) {
+				return ["AArch64","AMDGPU","ARM","AVR","BPF","Hexagon","Lanai","Mips","MSP430","NVPTX","PowerPC","RISCV","Sparc","SystemZ","WebAssembly","X86","XCore"];
```
Comment (Calrama) [inline source/llvm/config.d]: You reordered MSP430 and Mips. Semantically irrelevant, but it irked me, so I changed it ;)

File: source/llvm/config.d (hunk 2 above)
Comment (foerdi) [inline source/llvm/config.d]: Oh copy & paste error ... You have already merged this PR, should I open a new one?

Comment (Calrama) [inline source/llvm/config.d]: Nah, I've fixed it in master, no worries.

File: source/llvm/config.d (hunk 3 above)
Comment (foerdi) [inline source/llvm/config.d]: I wrote my own bash command to generate these lines of code and that has sorted the architectures alphabetically, next time I'll use your script :)

Comment (Calrama) [inline source/llvm/config.d]: What irked me was that there was a change from 3.9 to 4.0 :)
I have no issue with it being sorted alphabetically, but you'll have to do that for all of them then. I need consistency.

**Pull request merged**